filter/response.go

```diff
@@ -91,9 +92,76 @@ func (f *Filter) UpdateResponse(r *http.Response) error {
 		r.Header["Content-Length"] = []string{fmt.Sprint(buf.Len())}
 	}
 
+	if len(f.response.Header) > 0 {
+		r.Header = f.headerReplace(requestLog, r.Header, "response")
+	}
+
 	return nil
 }
 
+//UpdateRequest will be called back when the request is received by the proxy.
+func (f *Filter) UpdateRequest(r *http.Request) {
+	requestLog := f.log.WithFields(logrus.Fields{"url": r.URL.String(), "action": "request", "source": r.RemoteAddr})
+	requestLog.Debug("Request")
+
+	u, _ := url.Parse(f.url)
+	r.URL.Host = u.Host
+	r.Host = u.Host
+	r.URL.Scheme = "http"
+	data, err := httputil.DumpRequest(r, false)
+
+	if err != nil {
+		f.log.Error(fmt.Printf("Error"))
+	}
+
+	f.log.Debug(fmt.Sprintf("Request received\n %s", string(data)))
+
+	if r.Body != nil {
+		s, err := f.readBody(r.Body, r.Header)
+		if err != nil {
+			f.log.Fatal(err)
+		}
+
+		switch r.Header.Get("Content-Encoding") {
+		case "gzip":
+			w, _ := f.compress(s)
+
+			r.Body = ioutil.NopCloser(w)
+			r.Header["Content-Length"] = []string{fmt.Sprint(w.Len())}
+
+		default:
+			buf := bytes.NewBufferString(s)
+			r.Body = ioutil.NopCloser(buf)
+			r.Header["Content-Length"] = []string{fmt.Sprint(buf.Len())}
+		}
+	}
+
+	if len(f.request.Header) > 0 {
+		r.Header = f.headerReplace(requestLog, r.Header, "request")
+	}
+}
+
+func (f *Filter) headerReplace(log *logrus.Entry, h http.Header, a string) http.Header {
```

Comment:
Sorry, I have not looked to http.Header types, since it is an map and not a slice, no need to pass a pointer: [see this blog entry from one of Golang creator](https://dave.cheney.net/2017/04/30/if-a-map-isnt-a-reference-variable-what-is-it)